codice/ddf
# DDF-6242 Add timeout to CDM

#### What does this PR do?
- Creates a configurable timeout for files being processed by CDM. 
- Expired file names are stored in memory with their "last modified" date, and will only be reattempted once the file has been changed.
- Added Documentation and unit tests

#### Who is reviewing it? 
@aaronilovici 
@mdang8 
@clockard 

#### Select relevant component teams: 


#### Ask 2 committers to review/merge the PR and tag them here.
@brendan-hofmann
@clockard 

#### How should this be tested?
Steps to hero:
1. Build this branch and go through the regular install. In the console, set cdm logs to debug: `log:set DEBUG cdmLogger`
2. Set up a directory outside of dib and give DIB the proper permissions to access it.
3. In "etc/custom.system.properties", add the following line
    org.codice.ddf.catalog.content.monitor.expirationTime=10000
4. Set up a "Monitor In Place" CDM for the directory and add files to it to verify it ingests.
5. Add the *hanging-plugin* to the deploy directory (message for details). This will cause future files to hang during the ingest phase.
6. Add another file to the monitored directory. Wait 10 seconds
7. In the cdm logs (data/log/cdm.log) you should see that the recent file was expired:
      "Processing time limit expired for 0 files. Removing these files from processing: your_file.txt"
   Nothing should be logged after this.
8. Make a trivial change to the file added in step 6. The logs should try to process the file again, but should still fail.


#### Any background context you want to provide?

#### What are the relevant tickets?
Fixes: #____

#### Screenshots


#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
